zkevm-circuits/src/tx_circuit.rs

```diff
@@ -671,19 +724,223 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
+        });
+
+        meta.create_gate("block_num", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.condition(meta.query_advice(is_tag_block_num, Rotation::cur()), |cb| {
+                cb.require_equal(
+                    "block_num = tx_table.value if tag == BlockNum",
+                    meta.query_advice(tx_table.value, Rotation::cur()),
+                    meta.query_advice(block_num, Rotation::cur()),
+                );
+            });
+
+            cb.gate(meta.query_fixed(q_enable, Rotation::cur()))
+        });
+
+        // block num is the last row of each tx's fixed rows and since block num is
+        // copied to TX_LEN rows. The row at which tag = BlockNum and tx_id = i,
+        // its next row has tx_id = i+1. That is, we can use Rotation::next() to get next
+        // tx's all meta-infos (including block_num, tx_nonce, num_all_txs_acc, ...)
+        let block_num_unchanged = IsEqualChip::configure(
+            meta,
+            |meta| {
+                and::expr([
+                    meta.query_fixed(q_enable, Rotation::cur()),
+                    meta.query_advice(is_tag_block_num, Rotation::cur()),
+                ])
+            },
+            |meta| meta.query_advice(block_num, Rotation::next()),
+            |meta| meta.query_advice(block_num, Rotation::cur()),
+        );
+
+        meta.lookup(
+            "block_num' - block_num >= 1 if block_num changed for non-padding tx",
+            |meta| {
+                // Block nums like this [1, 3, 5, 4, 0] is rejected by this. But [1, 2, 3, 5, 0] is
+                // acceptable.
+                let lookup_condition = and::expr([
+                    // if next tx is padding, then its block num is 0 (i.e. may have block num
+                    // transition 5 -> 0)
+                    not::expr(meta.query_advice(is_padding_tx, Rotation::next())),
+                    // if tx table is full of real txs, then the last tx's next row is calldata.
+                    // we could have block num transition 5 -> 0.
+                    not::expr(meta.query_advice(is_calldata, Rotation::next())),
+                    not::expr(block_num_unchanged.expr()),
+                    meta.query_advice(is_tag_block_num, Rotation::cur()),
+                ]);
+
+                let block_num_diff = meta.query_advice(block_num, Rotation::next())
+                    - meta.query_advice(block_num, Rotation::cur());
+
+                vec![(lookup_condition * block_num_diff, u16_table.into())]
+            },
+        );
+
+        meta.create_gate("num_all_txs in a block", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+            let queue_index = tx_nonce;
+            // first tx in tx table
+            cb.condition(meta.query_fixed(q_second, Rotation::cur()), |cb| {
+                cb.require_equal(
+                    "first block's first tx's num_l1_msgs",
```

Comment:
```suggestion
                    "num_all_txs_acc = is_l1_msg ? queue_index - total_l1_popped_before + 1 : 1",
```

Reply:
fixed in https://github.com/scroll-tech/zkevm-circuits/pull/764/commits/bfd23e461fb503ff5fd8a74867829225a6e57fc8